feat: add liquidity provision subscription to GraphQL

guoguojin opened the pull request:
Closes #5888

guoguojin committed:
feat: add liquidity provision subscription to GraphQL

guoguojin committed:
Merge branch 'feat/5888-lp-commitments-gql-subscription' of github.com:vegaprotocol/vega into feat/5888-lp-commitments-gql-subscription

AnExsomnis commented on protos/sources/data-node/api/v2/trading_data.proto:
```diff
@@ -251,6 +251,8 @@ service TradingDataService {
     // Get a liquidity provision list for a given market using a cursor based pagination
     rpc ListLiquidityProvisions(ListLiquidityProvisionsRequest) returns (ListLiquidityProvisionsResponse);
 
+    rpc ObserveLiquidityProvisions(ObserveLiquidityProvisionsRequest) returns (stream ObserveLiquidityProvisionsResponse);
+
```
```suggestion
    // Get a liquidity provision subscription for a given market and party
    rpc ObserveLiquidityProvisions(ObserveLiquidityProvisionsRequest) returns (stream ObserveLiquidityProvisionsResponse);

```

guoguojin replied:
agreed, addressed https://github.com/vegaprotocol/vega/pull/6102/commits/4a45fd7642bdf0207c7b15f3866a8cd3bae8df79

AnExsomnis committed:
Merge branch 'develop' into feat/5888-lp-commitments-gql-subscription

guoguojin committed:
chore: address code review comments

guoguojin committed:
Merge branch 'develop' into feat/5888-lp-commitments-gql-subscription

guoguojin committed:
Merge branch 'feat/5888-lp-commitments-gql-subscription' of github.com:vegaprotocol/vega into feat/5888-lp-commitments-gql-subscription

jeremyletang commented on datanode/gateway/graphql/schema.graphql:
```diff
@@ -144,6 +144,14 @@ type Subscription {
+        partyId: ID
+        "the market ID to subscribe for, empty if all"
+        marketId: ID
+    ): [LiquidityProvision!]
```
This should be a specific type flattened like for other subscribtions (e.g: orders, trades etc).

guoguojin replied:
This is like this because of batching. The orders, trades subscriptions are declared:

```
  "Subscribe to orders updates"
  orders(
    "ID of the market from which you want orders updates"
    marketId: ID
    "ID of the party from which you want orders updates"
    partyId: ID
  ): [Order!]

  "Subscribe to the trades updates"
  trades(
    "ID of the market from which you want trades updates"
    marketId: ID
    "ID of the party from which you want trades updates"
    partyId: ID
  ): [TradeUpdate!]
```

The liquidity provisions events handler in the data-node also batches up the liquidity provision events before inserting to the database, so I've duplicated the subscription to send the batched updates. I can change it to stream them individually each time the batch is processed if you think this would be better?

guoguojin replied:
Sorry for the misunderstanding. I've addressed the requirement in https://github.com/vegaprotocol/vega/pull/6102/commits/04543f998b4fc67bb19ff712508089cfcf297e93

guoguojin committed:
chore: address review comments:

- Update documentation comments in protos to make things more clear
- Introduce LiquidityProvisionUpdate to flatten the subscription update
  response.

guoguojin commented on protos/sources/data-node/api/v2/trading_data.proto:
```diff
@@ -1503,6 +1505,17 @@ message LiquidityProvisionsConnection {
 }
 
+message ObserveLiquidityProvisionsRequest {
+    // The target market for the liquidity provision orders
```
Thanks, you are correct. I've updated the comments in https://github.com/vegaprotocol/vega/pull/6102/commits/04543f998b4fc67bb19ff712508089cfcf297e93

candida-d commented on datanode/gateway/graphql/schema.graphql:
```diff
@@ -3663,6 +3663,34 @@ type LiquidityProvision {
+  marketID: ID!
+  "Specified as a unit-less number that represents the amount of settlement asset of the market."
+  commitmentAmount: String!
+  "Nominated liquidity fee factor, which is an input to the calculation of maker fees on the market, as per setting fees and rewarding liquidity providers."
```
```suggestion
  "Nominated liquidity fee factor, which is an input to the calculation of liquidity fees on the market, as per setting fees and rewarding liquidity providers."
```

candida-d replied:
Does this come up anywhere else? I've seen this error crop up in FE things occasionally and I want to ensure it's totally fixed.

candida-d commented on datanode/gateway/graphql/schema.graphql:
```diff
@@ -3663,6 +3663,34 @@ type LiquidityProvision {
+  sells: [LiquidityOrderReference!]!
+  "a set of liquidity buy orders to meet the liquidity provision obligation."
+  buys: [LiquidityOrderReference!]!
+  "The version of this LiquidityProvision"
```
```suggestion
  "The version of this liquidity provision"
```

candida-d reviewed (commented):
@guoguojin Thanks for clarifying on the observers. I actually spotted one more thing I wanted to address, so I've also done a few tiny tidy-up suggestions on top - sorry to not get them all covered in one review.

guoguojin committed:
chore: additional corrections to graphql documentation per code review

guoguojin commented on datanode/gateway/graphql/schema.graphql:
```diff
@@ -3663,6 +3663,34 @@ type LiquidityProvision {
+  version: String!
+  "The current status of this liquidity provision"
+  status: LiquidityProvisionStatus!
+  "A reference for the orders created out of this Liquidity provision"
```
This and the others, along with the original documentation that also needed updating has been resolved with this commit [da260eb](https://github.com/vegaprotocol/vega/pull/6102/commits/da260ebfbab7edcf0b3a22d38511a6341f2cd36d)

guoguojin committed:
chore: resolve merge conflicts with develop
